### PR TITLE
fix: [PL-42268]: Add RBAC filtering on delegates when listing

### PR DESCRIPTION
**Pull request by VikasMaddukuriHarness**
## Describe your changes
If we have resourceGroup where it allows user to have view for only some delegates, we will filter out which delegates that are permitted and show up that in UI. Until now support was to view ALL or None delegates.

Testing:
1. Added UTs
2. Tested in devspaces
   - created 2 delegates and provided the permission to only one delegate in a delegate role and also created corresponding resource group. Created a user with that permission and checked the delegates list and verified only one (which is permitted) is allowed.
   - Tested the list delegates page and list delegate tags page which is called when delegate selectors are called in Sm creation step and verified only one which is permitted is being visible and only those tags are being visible in UI.
   
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55884)

### CODE REVIEW

**Review by Jimit09Gandhi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55884)* status: 6 file(s) to review. 1 change requests. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java` line 36 at r1](https://reviewable.io/reviews/harness/harness-core/55884#-Njzq55YCoesVFXDgq_Y:-Njzq55YCoesVFXDgq_Z:b-ucr9y0) ([raw file](https://github.com/harness/harness-core/blob/1b3fff27f5bf4085a57cf71f44bac3a6c1781dc7/420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java#L36)):*
> ```Java
> @OwnedBy(PL)
> public class DelegateRbacHelper {
>   @Inject private AccessControlClient accessControlClient;
> ```

Is this client privliege/Non-privileged? It should be Non-privileged
___
*[`420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java` line 57 at r1](https://reviewable.io/reviews/harness/harness-core/55884#-Njzrh2MC0V3SOIh_GBr:-Njzrh2MC0V3SOIh_GBs:b-u282ql) ([raw file](https://github.com/harness/harness-core/blob/1b3fff27f5bf4085a57cf71f44bac3a6c1781dc7/420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java#L57)):*
> ```Java
>             .collect(Collectors.toList());
>     AccessCheckResponseDTO accessCheckResponse = accessControlClient.checkForAccessOrThrow(permissionChecks);
> ```

Add UTs for this change.
___
*[`420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java` line 554 at r1](https://reviewable.io/reviews/harness/harness-core/55884#-NjzrbC8B5B4fS7hEp1w:-NjzrbC8B5B4fS7hEp1x:b-u282ql) ([raw file](https://github.com/harness/harness-core/blob/1b3fff27f5bf4085a57cf71f44bac3a6c1781dc7/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L554)):*
> ```Java
>     if (applyRbacFilter) {
>       delegateGroupIds = delegateRbacHelper.getPermittedIds(delegateGroupIds, accountId, orgId, projectId);
>     }
> ```

Add UTs for this change.

**Review by Jimit09Gandhi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55884)* status: 6 file(s) to review. 1 change requests. 0 of 2 approvals obtained. Waiting on team(s): del-team. 

___
*[`420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java` line 130 at r1](https://reviewable.io/reviews/harness/harness-core/55884#-Nk-Me1XBavHrzRBfrN_:-Nk-Me1XBavHrzRBfrNa:b-3i3h3l) ([raw file](https://github.com/harness/harness-core/blob/1b3fff27f5bf4085a57cf71f44bac3a6c1781dc7/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L130)):*
> ```Java
>             filterIdentifier, searchTerm, delegateFilterPropertiesDTO, pageRequest, false));
>       }
>     }
> ```

There are other delegate list APIs that needs to be changed as well.

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+public class DelegateRbacHelper {
+  @Inject private AccessControlClient accessControlClient;
+
+  public List<String> getPermittedIds(List<String> delegateGroupIds, String accountId, String orgId, String projectId) {
+    if (isEmpty(delegateGroupIds)) {
+      return null;
+    }
+
+    Map<EntityScopeInfo, List<String>> delegateGroupIdMap = delegateGroupIds.stream().collect(groupingBy(delegateGroupId
+        -> DelegateRbacHelper.getEntityScopeInfoFromDelegateGroupId(delegateGroupId, accountId, orgId, projectId)));
```
Comment (xingchi-jin) [inline 420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java:44]: you can invoke the getEntityScopeInfoFromDelegateGroupId without DelegateRbacHelper

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java [hunk 2]
```diff
@@ -548,6 +549,11 @@ public DelegateGroupListing listDelegateGroupDetailsV2(String accountId, String
     }
 
     List<String> delegateGroupIds = getDelegateGroupIds(accountId, orgId, projectId, filterProperties, searchTerm);
+    log.info("delegateGroupIds before filter: {}", delegateGroupIds);
```
Comment (xingchi-jin) [inline 420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java:552]: remove log

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.Map;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+@OwnedBy(PL)
+public class DelegateRbacHelper {
+  @Inject private AccessControlClient accessControlClient;
+
+  public List<String> getPermittedIds(List<String> delegateGroupIds, String accountId, String orgId, String projectId) {
```
Comment (xingchi-jin) [inline 420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java:38]: we can rename to "getViewPermittedDelegateGroupIds"

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        delegateGroupIds.stream()
+            .map(delegateGroupId
+                -> PermissionCheckDTO.builder()
+                       .permission(DELEGATE_VIEW_PERMISSION)
+                       .resourceIdentifier(delegateGroupId)
+                       .resourceScope(ResourceScope.of(accountId, orgId, projectId))
+                       .resourceType(DELEGATE_RESOURCE_TYPE)
+                       .build())
+            .collect(Collectors.toList());
+    AccessCheckResponseDTO accessCheckResponse = accessControlClient.checkForAccessOrThrow(permissionChecks);
```
Comment (xingchi-jin) [inline 420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java:56]: just a qq abt access control: when the principal is null, how does it check

**Review by xingchi-jin — APPROVED** [review]
LG with some nits

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java (hunk 4 above)
Comment (VikasMaddukuriHarness) [reply 420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java:67]: It gets the principal from the context of the thread. Correct me if i am wrong.

**Review by xingchi-jin — APPROVED** [review]

File: 420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java (hunk 4 above)
Comment (Jimit09Gandhi) [reply 420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java:68]: It derives principal from either bearer token/API key token.

**Review by Jimit09Gandhi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55884)* status: 6 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java` line 78 at r4](https://reviewable.io/reviews/harness/harness-core/55884#-NkUxe6h5CcfVb4JasuP:-NkUxe6h5CcfVb4JasuQ:bervp6a) ([raw file](https://github.com/harness/harness-core/blob/f0e6d5eddea743d99401c4381b7b75c22e899b73/420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java#L78)):*
> ```Java
>     }
>     Query<DelegateGroup> delegateGroupQueryToConvertIdentifiersToId =
>         persistence.createQuery(DelegateGroup.class).field(DelegateGroupKeys.identifier).in(permittedDelegateGroupIds);
> ```

Check for empty permittedDelegateGroupIds as query will throw exception.

**Review by VikasMaddukuriHarness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55884)* status: 6 file(s) to review. 1 change requests. 4. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java` line 36 at r1](https://reviewable.io/reviews/harness/harness-core/55884#-Njzq55YCoesVFXDgq_Y:-NkUyE7FA4jyPWOpiXIe:b-u93zvi) ([raw file](https://github.com/harness/harness-core/blob/1b3fff27f5bf4085a57cf71f44bac3a6c1781dc7/420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java#L36)):*
<details><summary><i>Previously, Jimit09Gandhi (Jimit Gandhi) wrote…</i></summary><blockquote>

Is this client privliege/Non-privileged? It should be Non-privileged
</blockquote></details>

it is non-priviliged onl
___
*[`420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java` line 38 at r2](https://reviewable.io/reviews/harness/harness-core/55884#gh-1407069743:-NkUyK5i4WBMVEVNDSPL:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9b13a019f22fce85d9381949317b8207bb2b22c9/420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java#L38)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

we can rename to "getViewPermittedDelegateGroupIds"
</blockquote></details>

Done.
___
*[`420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java` line 44 at r2](https://reviewable.io/reviews/harness/harness-core/55884#gh-1407063699:-NkUyKw59SZC8dvyx4Ng:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9b13a019f22fce85d9381949317b8207bb2b22c9/420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java#L44)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

you can invoke the getEntityScopeInfoFromDelegateGroupId without DelegateRbacHelper 
</blockquote></details>

Done.
___
*[`420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java` line 56 at r2](https://reviewable.io/reviews/harness/harness-core/55884#gh-1407071275:-NkUyMThAw2kCOOfdL30:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9b13a019f22fce85d9381949317b8207bb2b22c9/420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java#L56)):*
<details><summary><i>Previously, Jimit09Gandhi (Jimit Gandhi) wrote…</i></summary><blockquote>

It derives principal from either bearer token/API key token.
</blockquote></details>

Done.

**Review by Jimit09Gandhi — COMMENTED** [review]
> Reviewed 3 of 6 files at r1, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55884)* status: 6 file(s) to review. 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by VikasMaddukuriHarness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55884)* status: 15 file(s) to review. 8. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java` line 130 at r1](https://reviewable.io/reviews/harness/harness-core/55884#-Nk-Me1XBavHrzRBfrN_:-NknTH8G7OmpRyYV4ciY:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/1b3fff27f5bf4085a57cf71f44bac3a6c1781dc7/420-delegate-service/src/main/java/io/harness/delegate/resources/DelegateSetupResourceV2.java#L130)):*
<details><summary><i>Previously, Jimit09Gandhi (Jimit Gandhi) wrote…</i></summary><blockquote>

There are other delegate list APIs that needs to be changed as well.
</blockquote></details>

Done.
___
*[`420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java` line 57 at r1](https://reviewable.io/reviews/harness/harness-core/55884#-Njzrh2MC0V3SOIh_GBr:-NknTHtC4NaU3LqYFBOT:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/1b3fff27f5bf4085a57cf71f44bac3a6c1781dc7/420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java#L57)):*
<details><summary><i>Previously, Jimit09Gandhi (Jimit Gandhi) wrote…</i></summary><blockquote>

Add UTs for this change.
</blockquote></details>

Done.
___
*[`420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java` line 78 at r4](https://reviewable.io/reviews/harness/harness-core/55884#-NkUxe6h5CcfVb4JasuP:-NknTKKj0fGAFEtUIJju:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/f0e6d5eddea743d99401c4381b7b75c22e899b73/420-delegate-service/src/main/java/io/harness/service/impl/DelegateRbacHelper.java#L78)):*
<details><summary><i>Previously, Jimit09Gandhi (Jimit Gandhi) wrote…</i></summary><blockquote>

Check for empty permittedDelegateGroupIds as query will throw exception.
</blockquote></details>

Done.
___
*[`420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java` line 554 at r1](https://reviewable.io/reviews/harness/harness-core/55884#-NjzrbC8B5B4fS7hEp1w:-NknTM5Y7mKEgr7mT4zu:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/1b3fff27f5bf4085a57cf71f44bac3a6c1781dc7/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L554)):*
<details><summary><i>Previously, Jimit09Gandhi (Jimit Gandhi) wrote…</i></summary><blockquote>

Add UTs for this change.
</blockquote></details>

Done.
___
*[`420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java` line 552 at r2](https://reviewable.io/reviews/harness/harness-core/55884#gh-1407065791:-NknTLfODX76c0b3C5ff:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9b13a019f22fce85d9381949317b8207bb2b22c9/420-delegate-service/src/main/java/io/harness/service/impl/DelegateSetupServiceImpl.java#L552)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

remove log
</blockquote></details>

Done.

**Comment by VikasMaddukuriHarness** [comment]
trigger messagecheck

**Comment by VikasMaddukuriHarness** [comment]
trigger uts

**Review by Jimit09Gandhi — DISMISSED** [review]

**Comment by VikasMaddukuriHarness** [comment]
trigger ut5

**Comment by VikasMaddukuriHarness** [comment]
trigger uts

**Review by xingchi-jin — APPROVED** [review]

**Pull request merged**